zypp/ZYppFactory.cc

```diff
@@ -365,32 +365,41 @@ namespace zypp
       else if ( globalLock().zyppLocked() )
       {
 	bool failed = true;
-	const long LOCK_TIMEOUT = str::strtonum<long>( getenv( "ZYPP_LOCK_TIMEOUT" ) );
-	if ( LOCK_TIMEOUT > 0 )
+	const long LOCK_TIMEOUT = str::strtonum<long>( getenv( "ZYPP_LOCK_TIMEOUT" ) ); // <0 means forever
+	if ( LOCK_TIMEOUT != 0 )
 	{
-	  MIL << "Waiting whether pid " << globalLock().lockerPid() << " ends within $LOCK_TIMEOUT=" << LOCK_TIMEOUT << " sec." << endl;
-	  unsigned delay = 1;
-          Pathname procdir( Pathname("/proc")/str::numstring(globalLock().lockerPid()) );
-	  for ( long i = 0; i < LOCK_TIMEOUT; i += delay )
-	  {
-	    if ( PathInfo( procdir ).isDir() )	// wait for /proc/pid to disapear
-	      sleep( delay );
-	    else
+	  Date logwait = Date::now();
+	  Date giveup; /* 0 = forever */
+	  if ( LOCK_TIMEOUT > 0 ) {
+	    giveup = logwait+LOCK_TIMEOUT;
+	    MIL << "$ZYPP_LOCK_TIMEOUT=" << LOCK_TIMEOUT << " sec. Waiting for the zypp lock until " << giveup << endl;
+	  }
+	  else
+	    MIL << "$ZYPP_LOCK_TIMEOUT=" << LOCK_TIMEOUT << " sec. Waiting for the zypp lock..." << endl;
+
+	  unsigned delay = 0;
+	  do {
+	    if ( delay < 60 )
+	      delay += 1;
+	    else {
+	      Date now { Date::now() };
+	      if ( now - logwait > Date::day ) {
+		WAR << "$ZYPP_LOCK_TIMEOUT=" << LOCK_TIMEOUT << " sec. Another day has passed waiting for the zypp lock..." << endl;
+		logwait = now;
+	      }
+	    }
+	    sleep( delay );
 	    {
-	      MIL << "Retry after " << i << " sec." << endl;
+	      zypp::base::LogControl::TmpLineWriter shutUp;	// be quiet
 	      failed = globalLock().zyppLocked();
-	      if ( failed )
-	      {
-		// another proc locked faster. maybe it ends fast as well....
-		MIL << "Waiting whether pid " << globalLock().lockerPid() << " ends within " << (LOCK_TIMEOUT-i) << " sec." << endl;
-                procdir = Pathname( Pathname("/proc")/str::numstring(globalLock().lockerPid()) );
-	      }
-	      else
-	      {
-		MIL << "Finally got the lock!" << endl;
-		break;	// gotcha
-	      }
 	    }
+	  } while ( failed && ( not giveup || Date::now() <= giveup ) );
```

Comment:
Is the use of alternative operators intentional? I personally don't have a strong opinion about them, I'm just wondering.

Reply:
I think `not` is the only one I use sometimes. It may depend on the amount of python code I'm writing.